include/seqan3/range/view/kmer_hash.hpp

```diff
@@ -12,47 +12,616 @@
+     * Note that any shape is expected to start with a `1` and end with a `1`.
+     *
+     * ### Implementation detail
+     *
+     * To avoid dereferencing the sentinel when iterating, the shape_iterator computes the hash value up until
+     * the second to last position and performs the addition of the last position upon 
+     * access (\ref operator* and \ref operator[]).
+     */
+    template <typename rng_t>
+    class shape_iterator
+    {
+    private:
+        //!\brief The iterator type of the underlying range.
+        using it_t = std::ranges::iterator_t<rng_t>;
+        //!\brief The sentinel type of the underlying range.
+        using sentinel_t = std::ranges::sentinel_t<rng_t>;
+
+    public:
+        /*!\name Associated types
+         * \{
+         */
+        //!\brief Type for distances between iterators.
+        using difference_type = typename it_t::difference_type;
+        //!\brief Value type of this iterator.
+        using value_type = size_t;
+        //!\brief The pointer type;
+        using pointer = value_type *;
+        //!\brief Reference to `value_type`.
+        using reference = value_type;
+        //!\brief Reference to `value_type const`.
+        using const_reference = value_type const;
+        //!\brief Tag this class depending on which concept it_t models.
+        using iterator_category = std::conditional_t<std::ContiguousIterator<it_t>,
+                                                     typename std::random_access_iterator_tag,
+                                                     std::conditional_t<std::BidirectionalIterator<it_t>,
+                                                                        typename std::bidirectional_iterator_tag,
+                                                                        typename std::forward_iterator_tag>>;
+        //!\}
+
+        /*!\name Constructors, destructor and assignment
+         * \{
+         */
+        shape_iterator()                                   = default; //!< Defaulted.
+        shape_iterator(shape_iterator const &)             = default; //!< Defaulted.
+        shape_iterator(shape_iterator &&)                  = default; //!< Defaulted.
+        shape_iterator & operator=(shape_iterator const &) = default; //!< Defaulted.
+        shape_iterator & operator=(shape_iterator &&)      = default; //!< Defaulted.
+        ~shape_iterator()                                  = default; //!< Defaulted.
+        
+        /*!\brief Construct from a given iterator on the text and a seqan3::shape.
+        * /param[in] it_start Iterator pointing to the first position of the text.
+        * /param[in] s_ The seqan3::shape that determines which positions participate in hashing.
+        *
+        * ### Complexity
+        *
+        * Linear in size of shape.
+        */
+        shape_iterator(it_t it_start, kmer_hash_view const * v_) : 
```

Comment:
I would take by reference and take the address of inside, but I guess that's a matter of taste